agent/doctor/doctor.go

```diff
@@ -35,72 +30,92 @@ type Doctor struct {
+func (doc *Doctor) SetStatusReported(statusReported bool) {
+	doc.lock.Lock()
+	defer doc.lock.Unlock()
+
+	doc.statusReported = statusReported
+}
+
+// HasStatusBeenReported returns whether we have already sent the current
+// state of the healthchecks to the backend or not
+func (doc *Doctor) HasStatusBeenReported() bool {
+	doc.lock.RLock()
+	defer doc.lock.RUnlock()
+
+	return doc.statusReported
+}
+
+// AddHealthcheck adds a healthcheck to the list of healthchecks that the
+// doctor will run every time doctor.RunHealthchecks() is called
 func (doc *Doctor) AddHealthcheck(healthcheck Healthcheck) {
 	doc.lock.Lock()
 	defer doc.lock.Unlock()
 	doc.healthchecks = append(doc.healthchecks, healthcheck)
 }
 
+// RunHealthchecks runs every healthcheck that the doctor knows about and
+// returns a cumulative result; true if they all pass, false otherwise
 func (doc *Doctor) RunHealthchecks() bool {
-	doc.lock.RLock()
-	defer doc.lock.RUnlock()
+	doc.lock.Lock()
+	defer doc.lock.Unlock()
 	allChecksResult := []HealthcheckStatus{}
 
 	for _, healthcheck := range doc.healthchecks {
 		res := healthcheck.RunCheck()
 		seelog.Debugf("instance healthcheck result: %v", res)
 		allChecksResult = append(allChecksResult, res)
 	}
+
+	doc.statusReported = false
 	return doc.allRight(allChecksResult)
 }
 
-// GetPublishInstanceStatusRequest will get all healthcheck statuses and generate
-// a sendable PublishInstanceStatusRequest
-func (doc *Doctor) GetPublishInstanceStatusRequest() (*ecstcs.PublishInstanceStatusRequest, error) {
-	metadata := &ecstcs.InstanceStatusMetadata{
-		Cluster:           aws.String(doc.cluster),
-		ContainerInstance: aws.String(doc.containerInstanceArn),
-		RequestId:         aws.String(uuid.NewRandom().String()),
-	}
-	instanceStatuses := doc.getInstanceStatuses()
-
-	if instanceStatuses != nil {
-		return &ecstcs.PublishInstanceStatusRequest{
-			Metadata:  metadata,
-			Statuses:  instanceStatuses,
-			Timestamp: aws.Time(time.Now()),
-		}, nil
-	} else {
-		return nil, EmptyHealthcheckError
-	}
-}
+// GetHealthchecks returns a copy of list of healthchecks that the
+// doctor is holding internally.
+func (doc *Doctor) GetHealthchecks() *[]Healthcheck {
+	doc.lock.RLock()
+	defer doc.lock.RUnlock()
 
-func (doc *Doctor) getInstanceStatuses() []*ecstcs.InstanceStatus {
-	var instanceStatuses []*ecstcs.InstanceStatus
-	for _, healthcheck := range doc.healthchecks {
-		instanceStatus := &ecstcs.InstanceStatus{
-			LastStatusChange: aws.Time(healthcheck.GetStatusChangeTime()),
-			LastUpdated:      aws.Time(healthcheck.GetLastHealthcheckTime()),
-			Status:           aws.String(healthcheck.GetHealthcheckStatus().String()),
-			Type:             aws.String(healthcheck.GetHealthcheckType()),
-		}
-		instanceStatuses = append(instanceStatuses, instanceStatus)
-	}
-	return instanceStatuses
+	healthcheckCopy := make([]Healthcheck, len(doc.healthchecks))
+	copy(healthcheckCopy, doc.healthchecks)
+	return &healthcheckCopy
```

Comment:
Why are we making a copy and returning it instead of directly returning `doc.healthchecks`  ?

Reply:
Because we didn't want to pass the internal state of the Doctor around and have it get modified. This way, only one thing changes the internal state of the Doctor and that's when we run the healthchecks. All of the clients get their own copy and they are free to consume it/modify it/destroy it without affecting the internal state of the Doctor. And I was following how the TACS client already works for metrics.